Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,501 @@
+      wlog h : -1 < p
+      · simp only [show p = -1 by linarith, log_neg_eq_log, log_one, le_refl, mul_zero]
+      · nlinarith [log_neg_of_lt_zero hp h]
+    nlinarith [(log_pos (by linarith) : 0 < log (1 - p))]
+
+/-- Outside usual range of `binaryEntropy`, it is negative. This is due to `log x = log |x|` -/
+lemma binaryEntropy_neg_of_gt_one {p : ℝ} (hp : 1 < p) : binaryEntropy p < 0 := by
+  let x := p - 2⁻¹
+  rw [show p = 2⁻¹ + x by ring, binaryEntropy_two_inv_add]
+  have : 2⁻¹ - x < 0 := by ring_nf; linarith
+  exact binaryEntropy_neg_of_neg this
+
+lemma binaryEntropy_eq_zero {p : ℝ} : binaryEntropy p = 0 ↔ p = 0 ∨ p = 1 := by
+  constructor <;> intro h
+  · contrapose! h
+    obtain hp₀ | hp₀ := h.1.lt_or_lt
+    · exact (binaryEntropy_neg_of_neg hp₀).ne
+    · obtain hp₁ | hp₁ := h.2.lt_or_lt.symm
+      · exact (binaryEntropy_neg_of_gt_one hp₁).ne
+      · exact (binaryEntropy_pos hp₀ hp₁).ne'
+  · rw [binaryEntropy_apply]
+    cases h <;> simp only [log_one, mul_zero, sub_self, log_zero, neg_zero, log_zero, mul_zero,
+      sub_zero, log_one, sub_self, *]
+
+/-- For probability `p < 0.5`, `binaryEntropy p < log 2`. -/
+lemma binaryEntropy_lt_log2_of_lt_one_half {p : ℝ} (p_nonneg : 0 ≤ p) (p_lt : p < 1/2) :
+    binaryEntropy p < log 2 := by
+  -- Proof by concavity of log.
+  rw [binaryEntropy_apply]
+  rw [show -p * p.log = -(p * p.log) by ring]
+  by_cases pz : p = 0
+  · simp only [log_zero, mul_zero, neg_zero, sub_zero, log_one, sub_self, pz,
+      show 0 < log 2 by positivity]
+  · have invppos : 0 < 1/p := by positivity
+    have : 0 < 1 - p := by linarith -- used implicitly by tactics.
+    have sub1pinvpos : 0 < 1 / (1 - p) := by positivity
+    have logConcave := (strictConcaveOn_log_Ioi.right (x := 1/p) (y := 1/(1-p)))
+      (a := p) (b := 1-p) invppos sub1pinvpos (by norm_num; linarith) (by positivity)
+      (by linarith) (by norm_num)
+    have : p • (1 / p) + (1 - p) • (1 / (1 - p)) = 2 := by field_simp; norm_num
+    rw [this] at logConcave
+    have := calc -(p * log p) - (1 - p) * log (1 - p)
+      _ = p * (-log p) + (1 - p) * (-log (1 - p)) := by ring
+      _ = p * log (1/p) + (1 - p) * log (1 / (1 - p)) := by rw [← log_inv]; norm_num
+    rw [this]
+    exact logConcave
+
+lemma binaryEntropy_lt_log2_of_gt_half {p : ℝ} (h : 1/2 < p) (h2 : p ≤ 1) :
+    binaryEntropy p < log 2 := by
+  rw [← binaryEntropy_one_sub]
+  exact binaryEntropy_lt_log2_of_lt_one_half (by linarith) (by linarith)
+
+lemma binaryEntropy_eq_log2_iff_eq_half {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p = log 2 ↔ p = 1/2 := by
+  constructor <;> intro h
+  · by_cases h' : p < 1/2
+    · linarith [binaryEntropy_lt_log2_of_lt_one_half pge0 h']
+    · by_cases pgthalf : 1/2 < p
+      · linarith [binaryEntropy_lt_log2_of_gt_half pgthalf ple1]
+      · linarith
+  · simp only [one_div, binaryEntropy_two_inv, h]
+
+lemma binaryEntropy_le_log2 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p ≤ log 2 := by
```

Comment:
Can you be more consistent about whether the input to `qaryEntropy`/`binaryEntropy` is called `p` or `x`?

Reply:
Done.